### ISSUE TITLE
Moonsound/YMF278 issue: high notes sound bad (aliasing?)

### ISSUE THREAD

**Issue by MBilderbeek** [openMSX/openMSX#1400]
As Wolf reported.

To test: Run MBWave -> Select Waves -> Octave on C5 -> select Harmonica voice -> press U

The sound is bad, it's like aliasing. He tested also some other versions: 0.15 is the same, but 0.14 sounds clean(er).  With 0.14 the issue does show up when the first keyboard note is set to C6 (so at higher frequencies). Major changes in 0.15 were the great improvements authored by @ValleyBell . (Maybe he can help to analyze what's going on?)

The problem does not go away if I use the channel recorder [1], so it is in the actual sound output by the chip emulation at the native frequency (so, it's not a resampling issue).

[1] [channelrecorded_u.zip](https://github.com/openMSX/openMSX/files/7918926/channelrecorded_u.zip)
This is a small recording of the test case with different notes using Q W E R T Y U I O P [harmonica.avi.zip](https://github.com/openMSX/openMSX/files/7918925/harmonica.avi.zip) The issue is mostly present from U onwards.

**Comment by ValleyBell** [comment]
You don't happen to have a hardware recording of that, do you?

I'd like to rule out the possibility of this being a flaw with the actual chip.

**Comment by MBilderbeek** [comment]
@ValleyBell A friend of mine (Meits) checked on real hardware and there it did sound clean. I've requested him to record it, so I can attach it to this issue. 

FYI, how it sounds in openMSX 0.14.0 is attached here:
[openmsx0003.wav.zip](https://github.com/openMSX/openMSX/files/7919124/openmsx0003.wav.zip)
And this is the source (requires MSX-DOS2):
[mbwave.zip](https://github.com/openMSX/openMSX/files/7919123/mbwave.zip)

**Comment by MBilderbeek** [comment]
I made a VGM file for the test case provided here, so that it's easy to check for @ValleyBell whether his implementation has the same issue/difference.
[music0007.vgm.zip](https://github.com/openMSX/openMSX/files/7919341/music0007.vgm.zip)

**Comment by MBilderbeek** [comment]
@grauw Could you make a recording of this on real hardware, just to be sure?

**Comment by ValleyBell** [comment]
I can confirm that the aliasing got worse in libvgm/VGMPlay as well.
In the 2017 version (with the old core) there is almost no aliasing, but it is very audible with the rewritten OPL4 emulation.

**Comment by m9710797** [comment]
I did a bisect, and I can confirm that revision 6453d5c606f9b12691f6b22831e233d54dfa12c2 introduced the regression. (The revision right before still sounds fine, this revision shows the problem).
Unfortunately it's a rather large change, and it's not clear which of the many changes might be responsible.

@ValleyBell: Do you have any idea. I think the next step could be to (manually) bisect within this patch itself. Any suggestions for where to start? Or maybe you could do some experiments as well?

**Comment by ValleyBell** [comment]
I'll have a deeper look when I get the time.

The first spot where I'd look is the sample looping code: https://github.com/openMSX/openMSX/blob/aa0f2b8867af8976f4ac5953d66428d7f63516f5/src/sound/YMF278.cc#L536
The code for interpolating between samples (line 504) was not changed.
If that isn't it, then maybe it is related to the LFOs.

**Comment by MBilderbeek** [comment]
@ValleyBell @m9710797 Any news on this?

**Comment by grauw** [comment]
The issue is indeed introduced by the sample looping change [here](https://github.com/openMSX/openMSX/commit/6453d5c606f9b12691f6b22831e233d54dfa12c2#diff-0c536ff5ab1c92874b9384358672be0a64d4c9e43aa2ceebcffe94161f09e6bcR502).

If I revert these lines in `libvgm`, the test VGM recorded by Manual plays normally:

```diff
diff --git a/emu/cores/ymf278b.c b/emu/cores/ymf278b.c
index 8410a71..8cbb783 100644
--- a/emu/cores/ymf278b.c
+++ b/emu/cores/ymf278b.c
@@ -816,17 +816,15 @@ static void ymf278b_pcm_update(void *info, UINT32 samples, DEV_SMPL** outputs)
                             : sl->step;
                        sl->stepptr += step;
 
-                       // If there is a 4-sample loop and you advance 12 samples per step,
-                       // it may exceed the end offset.
-                       // This is abused by the "Lizard Star" song to generate noise at 0:52. -Valley Bell
-                       if (sl->stepptr >= 0x10000)
+                       while (sl->stepptr >= 0x10000)
                        {
+                               sl->stepptr -= 0x10000;
                                sl->sample1 = sl->sample2;
                                sl->sample2 = ymf278b_getSample(chip, sl);
-                               sl->pos += (sl->stepptr >> 16);
-                               sl->stepptr &= 0xFFFF;
-                               if ((UINT32)sl->pos + sl->endaddr >= 0x10000)   // check position >= (negated) end address
-                                       sl->pos = sl->pos + sl->endaddr + sl->loopaddr; // This is how the actual chip does it.
+                               if (sl->pos >= sl->endaddr)
+                                       sl->pos = sl->pos - sl->endaddr + sl->loopaddr;
+                               else
+                                       sl->pos ++;
                        }
                }
                ymf278b_advance(chip);
@@ -931,9 +929,10 @@ static void ymf278b_C_w(YMF278BChip* chip, UINT8 reg, UINT8 data)
                        }
                        
                        slot->bits = (buf[0] & 0xC0) >> 6;
-                       slot->startaddr = buf[2] | (buf[1] << 8) | ((buf[0] & 0x3F) << 16);
-                       slot->loopaddr = buf[4] | (buf[3] << 8);
-                       slot->endaddr  = buf[6] | (buf[5] << 8);
+                       slot->startaddr = buf[2] | (buf[1] << 8) |
+                                         ((buf[0] & 0x3F) << 16);
+                       slot->loopaddr = buf[4] + (buf[3] << 8);
+                       slot->endaddr  = ((buf[6] + (buf[5] << 8)) ^ 0xFFFF);
                        for (i = 7; i < 12; ++i)
                        {
                                // Verified on real YMF278:
```

Of course, [Lizard Star](https://www.msx.org/downloads/mrc-challenges/game-music-cover-challenge/salamander-starfield-omega) no longer plays the noise.

**Comment by grauw** [comment]
The problem is that after the looping change it no longer correctly interpolates between samples when the step size is greater than 1.

With this change both the test and Lizard Star sound like they’re supposed to:

```diff
diff --git a/emu/cores/ymf278b.c b/emu/cores/ymf278b.c
index 8410a71..300594f 100644
--- a/emu/cores/ymf278b.c
+++ b/emu/cores/ymf278b.c
@@ -821,9 +821,11 @@ static void ymf278b_pcm_update(void *info, UINT32 samples, DEV_SMPL** outputs)
                        // This is abused by the "Lizard Star" song to generate noise at 0:52. -Valley Bell
                        if (sl->stepptr >= 0x10000)
                        {
-                               sl->sample1 = sl->sample2;
-                               sl->sample2 = ymf278b_getSample(chip, sl);
-                               sl->pos += (sl->stepptr >> 16);
+                               for (UINT16 steps = sl->stepptr >> 16; steps > 0; --steps) {
+                                       sl->sample1 = sl->sample2;
+                                       sl->sample2 = ymf278b_getSample(chip, sl);
+                                       sl->pos++;
+                               }
                                sl->stepptr &= 0xFFFF;
                                if ((UINT32)sl->pos + sl->endaddr >= 0x10000)   // check position >= (negated) end address
                                        sl->pos = sl->pos + sl->endaddr + sl->loopaddr; // This is how the actual chip does it.
```

However I don’t think this is the fully correct solution yet. Let me think about it a bit longer, if this is really what the OPL4 would do.

Because effectively this means the OPL4 always does two memory reads instead of one. Then might as well get rid of the `sample1` and `sample2` queue. It was already weird that it sometimes would do two reads (key on), sometimes one and sometimes zero.

Additionally, although it does sound ok to me now, this does potentially read past the end point, while the old code did not. I wonder if the OPL4 takes that into account.

**Comment by grauw** [comment]
Alternate implementation based on the considerations above:

```diff
diff --git a/emu/cores/ymf278b.c b/emu/cores/ymf278b.c
index 8410a71..d5240db 100644
--- a/emu/cores/ymf278b.c
+++ b/emu/cores/ymf278b.c
@@ -173,8 +173,7 @@ typedef struct
        UINT32 step;    // fixed-point frequency step
                                        // invariant: step == calcStep(OCT, FN)
        UINT32 stepptr; // fixed-point pointer into the sample
-       UINT16 pos;
-       INT16 sample1, sample2;
+       UINT16 pos, pos2;
 
        INT16 env_vol;
 
@@ -457,7 +456,7 @@ static void ymf278b_slot_reset(YMF278BSlot* slot)
        slot->state = EG_OFF;
 
        // not strictly needed, but avoid UMR on savestate
-       slot->pos = slot->sample1 = slot->sample2 = 0;
+       slot->pos = slot->pos2 = 0;
 }
 
 INLINE UINT8 ymf278b_slot_compute_rate(YMF278BSlot* slot, int val)
@@ -694,7 +693,7 @@ static void ymf278b_advance(YMF278BChip* chip)
        }
 }
 
-INLINE INT16 ymf278b_getSample(YMF278BChip* chip, YMF278BSlot* op)
+INLINE INT16 ymf278b_getSample(YMF278BChip* chip, YMF278BSlot* op, UINT16 pos)
 {
        // TODO How does this behave when R#2 bit 0 = 1?
        //      As-if read returns 0xff? (Like for CPU memory reads.) Or is
@@ -706,19 +705,19 @@ INLINE INT16 ymf278b_getSample(YMF278BChip* chip, YMF278BSlot* op)
        {
        case 0:
                // 8 bit
-               sample = ymf278b_readMem(chip, op->startaddr + op->pos) << 8;
+               sample = ymf278b_readMem(chip, op->startaddr + pos) << 8;
                break;
        case 1:
                // 12 bit
-               addr = op->startaddr + ((op->pos / 2) * 3);
-               if (op->pos & 1)
+               addr = op->startaddr + ((pos / 2) * 3);
+               if (pos & 1)
                        sample = (ymf278b_readMem(chip, addr + 2) << 8) | ((ymf278b_readMem(chip, addr + 1) & 0xF0) << 0);
                else
                        sample = (ymf278b_readMem(chip, addr + 0) << 8) | ((ymf278b_readMem(chip, addr + 1) & 0x0F) << 4);
                break;
        case 2:
                // 16 bit
-               addr = op->startaddr + (op->pos * 2);
+               addr = op->startaddr + (pos * 2);
                sample = (ymf278b_readMem(chip, addr + 0) << 8) | ymf278b_readMem(chip, addr + 1);
                break;
        default:
@@ -781,8 +780,8 @@ static void ymf278b_pcm_update(void *info, UINT32 samples, DEV_SMPL** outputs)
                                continue;
                        }
 
-                       sample = (sl->sample1 * (0x10000 - sl->stepptr) +
-                                 sl->sample2 * sl->stepptr) >> 16;
+                       sample = (ymf278b_getSample(chip, sl, sl->pos) * (0x10000 - sl->stepptr) +
+                                 ymf278b_getSample(chip, sl, sl->pos2) * sl->stepptr) >> 16;
                        
                        // TL levels are 00..FF internally (TL register value 7F is mapped to TL level FF)
                        // Envelope levels have 4x the resolution (000..3FF)
@@ -821,12 +820,13 @@ static void ymf278b_pcm_update(void *info, UINT32 samples, DEV_SMPL** outputs)
                        // This is abused by the "Lizard Star" song to generate noise at 0:52. -Valley Bell
                        if (sl->stepptr >= 0x10000)
                        {
-                               sl->sample1 = sl->sample2;
-                               sl->sample2 = ymf278b_getSample(chip, sl);
                                sl->pos += (sl->stepptr >> 16);
+                               sl->pos2 += (sl->stepptr >> 16);
                                sl->stepptr &= 0xFFFF;
                                if ((UINT32)sl->pos + sl->endaddr >= 0x10000)   // check position >= (negated) end address
                                        sl->pos = sl->pos + sl->endaddr + sl->loopaddr; // This is how the actual chip does it.
+                               if ((UINT32)sl->pos2 + sl->endaddr >= 0x10000)  // check position >= (negated) end address
+                                       sl->pos2 = sl->pos2 + sl->endaddr + sl->loopaddr;       // This is how the actual chip does it.
                        }
                }
                ymf278b_advance(chip);
@@ -850,9 +850,7 @@ INLINE void ymf278b_keyOnHelper(YMF278BChip* chip, YMF278BSlot* slot)
        }
        slot->stepptr = 0;
        slot->pos = 0;
-       slot->sample1 = ymf278b_getSample(chip, slot);
-       slot->pos = 1;
-       slot->sample2 = ymf278b_getSample(chip, slot);
+       slot->pos2 = 1;
 }
 
 static void ymf278b_A_w(YMF278BChip *chip, UINT8 reg, UINT8 data)
```

This gets rid of sample1 and sample2, and instead just interpolates between two adjacent memory reads for each sample. It tracks two position counters with an offset of 1 to avoid reading past the end point. In terms of internal logic, the OPL4 could have a read-increment-loop circuit that’s executed twice, once for each read.

I think this is the most likely representation of OPL4 internals. What do you guys think?

**Comment by grauw** [comment]
In the mean while I got an openMSX patch ready along the lines of the change above, with some clean-ups and serialisation. Will do a pull request tomorrow, and for libvgm afterwards.

**Comment by grauw** [comment]
For reference, ValleyBell’s review comments:

> 17:21 ValleyBell: I just wonder.... shouldn't it be possible to test how the interpolation works?
> 17:22 ValleyBell: maybe using a special hand-crafted sample
> 17:25 grauw: Do you have suggestions on what specific parts to test further?
> 17:25 grauw: In principle I think the test case already demonstrates the issue fairly effectively, and this was the only way I could come up with which fixed both it and kept the Lizard Star effect functional.
> 17:26 grauw: Logically it makes sense to me that interpolation should happen between adjacent samples, position is a fixed point index and should not interpolate between nonadjacent samples.
> 17:26 grauw: Tho of course there is the loop case where they’re not adjacent (physically).
> 17:31 grauw: But my reasoning for just assuming the OPL4 handles that properly, is because otherwise interpolating into undefined memory at the end point would be such a common event.
> 17:31 grauw: Even for low increments it’d be almost unavoidable to overshoot.
> 17:34 grauw: Btw the last three notes in the test still do have some mild aliasing, which I can also hear on the real thing.
> 17:34 ValleyBell: yeah, that is to be expected
> 17:35 grauw: I think so
> 17:35 ValleyBell: For testing the interpolator, I thought of something like this:
> 17:35 ValleyBell: a sample: 0 +1 -1 +2 -2 +3 -3 +4 -4 etc
> 17:35 ValleyBell: and setting playback speed to "every 2nd byte"
> 17:36 ValleyBell: or "every 1.5th byte"
> 17:37 ValleyBell: That probably won't sound pleasant, but it should give clues as to what bytes the chip uses for the interpolation.
> 17:39 ValleyBell: Anyway: In general I think your fix looks good and it makes sense this way.
> 17:41 ValleyBell: I don't think that the chip keeps two separate sample positions though.
> 17:46 grauw: I’ll think about if I can come up with some additional tests.
> 17:47 grauw: Re. two positions, it seemed to me the simplest possible implementation without introducing either loops or many conditionals and variable number of reads.
> 17:48 ValleyBell: yeah, I can't really think of anything better either
> 18:20 grauw: Wouterv suggested to replace getSample(sl, sl.pos2) with getSample(sl, nextPos(sl.pos, 1)), then only one position needs to be tracked.
> 18:22 grauw: Not sure if that’s how the OPL4 would work internally since it is just a slightly less consistent path, but for a C++ implementation it doesn’t matter anyway as long as the result is the same.

**Comment by grauw** [comment]
Small slightly (un)related addendum, just to record my findings somewhere:

As I was looking at the implementation of `nextPos` I noticed that if the end address is set to its maximum value (0, meaning 64k samples), it will start from the beginning and ignore the loop address. This because `pos` is a 16-bit value which wraps around, and adding 0 to it will never exceed FFFFH.

I can confirm that the real OPL4 matches this behaviour. [This test VGM](https://github.com/openMSX/openMSX/files/8353765/testopl4loop.vgm.zip) has its loop address set half-way, however the OPL4 plays it from the start every time. If I set the end address to FFFFH (1), it does loop from the half-way point.
